Adds Orange Pi GPIO platform

pascallj opened the pull request:
## Description:

This component adds support for GPIO on Orange Pi boards. It uses `rpi_gpio` as base without support for port debouncing or pull-up resistors. These features are (currently) not supported by the OPi.GPIO library. This library uses SYSFS to access the GPIOs and therefore need some additional steps to get working in HASS (see documentation).

As there exist different pinouts for different boards, an additional `pinmode` attribute has been added to support all boards supported by the library (see documentation).

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#9072

## Example entry for `configuration.yaml` (if applicable):
```yaml
# Example configuration.yaml entry
binary_sensor:
  - platform: orangepi_gpio
    pinmode: pc
    ports:
      11: PIR Office
      12: PIR Bedroom
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard/__init__.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard/__init__.py#L23

pascallj committed:
Adds Orange Pi GPIO platform

homeassistant commented:
Hi @pascallj, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#22541).

Once you do that we will be able to review and accept this pull request.

Thanks!

fabaff commented:
Successor of #19732.

rohankapoorcom commented on homeassistant/components/orangepi_gpio/binary_sensor.py:
```diff
@@ -0,0 +1,86 @@
+CONF_PINMODE = 'pinmode'
+CONF_PORTS = 'ports'
+
+PINMODES = ['pc', 'zeroplus', 'zeroplus2', 'deo', 'neocore2']
```
Pull these to a shared constant rather than redefining for each platform.

rohankapoorcom commented on homeassistant/components/orangepi_gpio/binary_sensor.py:
```diff
@@ -0,0 +1,86 @@
+    """Represent a binary sensor that uses Orange Pi GPIO."""
+
+    def __init__(self, name, port, invert_logic):
+        """Initialize the RPi binary sensor."""
```
Stale docstring

rohankapoorcom commented on homeassistant/components/orangepi_gpio/cover.py:
```diff
@@ -0,0 +1,115 @@
+_LOGGER = logging.getLogger(__name__)
+
+CONF_COVERS = 'covers'
+CONF_PINMODE = 'pinmode'
```
Pull all shared constants to a single consts file.

rohankapoorcom commented on homeassistant/components/orangepi_gpio/cover.py:
```diff
@@ -0,0 +1,115 @@
+
+def setup_platform(hass, config, add_entities, discovery_info=None):
+    """Set up the Orange Pi cover platform."""
+    pinmode = config.get('pinmode')
```
Use the constants defined
```suggestion
    pinmode = config[CONF_PINMODE]
```

rohankapoorcom commented on homeassistant/components/orangepi_gpio/cover.py:
```diff
@@ -0,0 +1,115 @@
+    pinmode = config.get('pinmode')
+    orangepi_gpio.setup_mode(pinmode)
+
+    relay_time = config.get(CONF_RELAY_TIME)
```
Since these all have default values, use the [] key access syntax

rohankapoorcom commented on homeassistant/components/orangepi_gpio/cover.py:
```diff
@@ -0,0 +1,115 @@
+        self._invert_relay = invert_relay
+        orangepi_gpio.setup_output(self._relay_pin)
+        orangepi_gpio.setup_input(self._state_pin)
+        orangepi_gpio.write_output(self._relay_pin,
```
It's considered best practice to avoid doing other things in the Entity initialization. Is there a better place to move `write_output`?

pascallj replied:
This was just inherited from rpi_gpio. I was able to trace this code all the way back to the original ['Garage Door' project by andrewshilliday](https://github.com/andrewshilliday/garage-door-controller). But I couldn't find a reason why the output is written here. Possible solution to just remove it?

rohankapoorcom commented on homeassistant/components/orangepi_gpio/switch.py:
```diff
@@ -0,0 +1,86 @@
+CONF_PINMODE = 'pinmode'
+CONF_INVERT_LOGIC = 'invert_logic'
+
+PINMODES = ['pc', 'zeroplus', 'zeroplus2', 'deo', 'neocore2']
```
Same comment on pulling shared constants out of this file.

rohankapoorcom commented on homeassistant/components/orangepi_gpio/switch.py:
```diff
@@ -0,0 +1,86 @@
+
+DEFAULT_INVERT_LOGIC = False
+
+_SWITCHES_SCHEMA = vol.Schema({
```
This Schema is the same as the sensor schema, let's use one of them from a const file instead.

pascallj replied:
Not sure if the solution I provided is desirable: import parts of the binary sensor into the switch and re-use them. However it does reduce duplicated code.

rohankapoorcom replied:
We should not import form the binary sensor into the switch. Instead move the schema to a shared const.py file or the init.py file and import it in both the binary sensor and the switch.

rohankapoorcom commented on homeassistant/components/orangepi_gpio/switch.py:
```diff
@@ -0,0 +1,86 @@
+    cv.positive_int: cv.string,
+})
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
```
This Schema is identical to the one for the binary_sensor. Let's use one of them from a const file instead

rohankapoorcom commented on homeassistant/components/orangepi_gpio/switch.py:
```diff
@@ -0,0 +1,86 @@
+
+def setup_platform(hass, config, add_entities, discovery_info=None):
+    """Set up the Orange Pi GPIO devices."""
+    pinmode = config.get('pinmode')
```
Use the constants defined
```suggestion
    pinmode = config[CONF_PINMODE]
```

rohankapoorcom reviewed (changes requested):
I've finished a first pass review. It's a great start :)

Please see the comments with detailed changes requested.

pascallj commented:
Thanks for your first review, I appreciate your optimisim :).

awarecan commented:
It is harsh time for new integration as we are doing refactoring, we are now requesting all components has to have a [manifest file](https://github.com/home-assistant/architecture/issues/179).

Please rebase you branch against current `dev` branch, and create a `manifest.json` file under your component's folder.

You may need to do more rebase in the next few days as we are still in the progress of refactoring.

pascallj committed:
Add manifest.json

pascallj committed:
Remove cover platform

pascallj commented:
As I could not work out why the output is written in the initialization stage of the cover platform, I removed the platform for now so we can get this through. Platform can be added at a later stage.

Also I rebased a couple days ago but as far as I can see it is still up-to-date with the refactoring process.

pascallj committed:
Apply requested changes

pascallj committed:
Remove switch platform

pascallj commented on homeassistant/components/orangepi_gpio/binary_sensor.py:
```diff
@@ -0,0 +1,70 @@
+
+def setup_platform(hass, config, add_entities, discovery_info=None):
+    """Set up the Orange Pi GPIO devices."""
+    pinmode = config.get(CONF_PINMODE)
```
Sorry you mentioned that before, changes must be lost somewhere.

pascallj commented:
The changes you requested for the switch platform have the same issue as the cover platform. I assume the fix is simple, but needs further testing before I can verify that. Therefore I will limit this PR to just one platform, the binary_sensor, for now (as instructed by the component checklist).

pascallj committed:
Update CODEOWNERS

pascallj commented:
Any other changes I have to make to this PR?

rohankapoorcom commented on homeassistant/components/orangepi_gpio/__init__.py:
```diff
@@ -0,0 +1,83 @@
+from homeassistant.const import (
+    EVENT_HOMEASSISTANT_START, EVENT_HOMEASSISTANT_STOP)
+
+REQUIREMENTS = ['OPi.GPIO==0.3.6']
```
This is no longer needed

rohankapoorcom commented on homeassistant/components/orangepi_gpio/binary_sensor.py:
```diff
@@ -0,0 +1,70 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+DEPENDENCIES = ['orangepi_gpio']
```
This should also not be needed anymore.

rohankapoorcom commented:
Hi @pascallj - thanks for your patience as we went through a major refactor during the process of you writing your PR. I’ve marked two lines that need to be removed, but otherwise this looks good to me!

pascallj committed:
Remove obsolete dependecies/requirements

pascallj commented:
Thank you, I removed those lines. Waiting for the tests to pass :)

MartinHjelmare commented on homeassistant/components/orangepi_gpio/binary_sensor.py:
```diff
@@ -0,0 +1,68 @@
+            self._state = orangepi_gpio.read_input(self._port)
+            self.schedule_update_ha_state()
+
+        orangepi_gpio.edge_detect(self._port, read_gpio)
```
State update callbacks should be registered in `async_added_to_hass` entity coroutine method.

pascallj replied:
Better late than never: #27958

MartinHjelmare commented on homeassistant/components/orangepi_gpio/binary_sensor.py:
```diff
@@ -0,0 +1,68 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(PORT_SCHEMA)
```
Since we know that more platforms are likely incoming, it would be better to have all config under the integrating component section, and set up platforms via discovery helper. See the netgear_lte component for example.

MartinHjelmare commented on homeassistant/components/orangepi_gpio/binary_sensor.py:
```diff
@@ -0,0 +1,68 @@
+"""Support for binary sensor using Orange Pi GPIO."""
+import logging
+
+from homeassistant.components import orangepi_gpio
```
Integration package imports should always be relative.
```py
from . import setup_input, setup_mode, etc
```

pascallj replied:
#27958

MartinHjelmare commented on homeassistant/components/orangepi_gpio/binary_sensor.py:
```diff
@@ -0,0 +1,68 @@
+        self._invert_logic = invert_logic
+        self._state = None
+
+        orangepi_gpio.setup_input(self._port)
```
We try to avoid side effects in init method.

pascallj replied:
What side effects are you referring to?

MartinHjelmare replied:
We're setting the port as an input channel, correct? So we're doing something besides returning a value.

pascallj replied:
True, but in my opinion this still belongs to the term initialization. But I will compare against other components to see how it should be done instead.

MartinHjelmare replied:
Either move it to `async_added_to_hass` and schedule the call on the executor thread pool, or move it to `setup_platform` before creating the entity.

pascallj replied:
#27958

MartinHjelmare commented on homeassistant/components/orangepi_gpio/__init__.py:
```diff
@@ -0,0 +1,81 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+CONF_PINMODE = 'pinmode'
```
```py
CONF_PIN_MODE = 'pin_mode'
```
Please separate words by underscore.

pascallj replied:
https://github.com/home-assistant/home-assistant/pull/23231

MartinHjelmare reviewed (commented):
Please open a new PR where we can address the comments.
